src/Dialect/ONNX/ONNXOps/ShapeHelper.cpp

```diff
@@ -810,6 +810,55 @@ void resetTypesShapeToQuestionmarks(Operation *op) {
     resetTypeShapeToQuestionmarks(op->getResult(i));
 }
 
+//===----------------------------------------------------------------------===//
+// ONNX Custom Op Shape Helper
+//===----------------------------------------------------------------------===//
+
+ONNXCustomOpShapeHelper::ONNXCustomOpShapeHelper(Operation *op,
```

Comment:
This code does not respect the protocol, and will make this shape helper not compatible with other analysis phases. The protocol is as follows:

constructor -> save all the info you need from the op; the constructor cannot have any additional parameters
computeShape -> perform the computation, cannot accept any additional parameters.

In the current form, you compute the shape inside the constructor... and provide no new function for computeShape.

Here is the refactoring that you need to perform;
save `pattern` as you do
update the list of operands. your ToFix comment is correct; if you update the list of operands, you need to see how this is done in the constructor of OnnxShapeHelper (ShapeHelper.cpp, line 97).

then provide a custom `computeShape` which will use `pattern` and the (possibly pruned) list of operands, and will use one of the two shapes depending on pattern.

Reply:
1. ShapeHelper::privateOperandsCache can be used to replace the ONNXCustomOpShapeHelper::operandsVector. However, they are implementation support of their own class. To combine them, ShapeHelper::privateOperandsCache has to be read/written. And ShapeHelper::operands will become invalid when the privateOeprandsCache is cleared to construct the new Ops. I'd like to keep the code as it is. If you insist, I will change it.
2. computeShape is already there.

Reply:
Added setOperands to use privateOperandsCache
computeShape is already there.